[RFC] Allow `=` in parameter values.

geier opened the pull request:
Some parameter values (e.g., BASE64 encoded binary data often ends with one or two equal signs) may contain an equal sign (`=`). The current implementation splits key-value pairs at all equal signs, which leads to errors. Especially icalendar files generated by Apple's software often feature BASE64 encoded binary data in parameter values.

This patch introduces a new parameter `maxsplit` to icalendar.parser.q_split() which works similar as python's string.split(sep, maxsplit) which we then use to split parameter key-value pairs only at the first equal sign.

This patch fixes #197.

digsim commented:
Regarding the test introduced in #204 one could argue to remove this test now. Its only purpose was to pinpoint a problem, which is getting fixed by this commit making this test obsolete. However, I would rather argue that the test should stay. I would therefore suggest to change line 20 into
`self.assertEqual(len(event.errors), 0, 'Got too many errors'`
and remove lines 21 and 22. 

This second approach would then verify that:
1. icalendar is able to parse a file containing binary information (**already possible without this patch**) and
2. icalendar can do this without producing any errors (**new with this patch**).

geier commented:
I have modified the old test and added a new one that tests the fix introduced in #204.

geier commented:
Looks like I forgot to push those changes (done now). The failing test is pypy3, which we currently do not support, I have opened #206 to deal with that.

untitaker commented on src/icalendar/tests/test_icalendar.py:
```diff
@@ -247,7 +275,50 @@ def test_q_split(self):
+            for event in cal.walk('vevent'):
+                self.assertEqual(len(event.errors), 0, 'Got too many errors')
+
+        except UnicodeEncodeError as e:
```
Why catch the error at all?

geier replied:
Honestly: no idea, I've just re-used the old test, but sure, we can remove it.

geier commented:
I removed those try/excepts.

untitaker commented:
LGTM

On Wed, Jan 04, 2017 at 05:20:34PM -0800, Christian Geier wrote:
> I removed those try/excepts.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/collective/icalendar/pull/207#issuecomment-270537032

digsim commented:
That should be ok. Back then, the except was used to let the test fail.

geier commented:
I'd like to get someone else's review on this. @stlaz, @thet or @regebro, if you find the time, I'd really appreciate your input.

stlaz commented:
Putting it on my TODO list

regebro commented:
"=" is a SAFECHAR and should be allowed in unquoted values, so this is indeed a bug.
Only comment I have is that passing in maxsplit=0 doesn't return the expected result (no splits). It's a sufficiently silly case that you can choose not to support it.

untitaker commented:
Apparently I can push commits to this PR, so I fixed this!

On Thu, Jan 19, 2017 at 03:03:28AM -0800, Lennart Regebro wrote:
> "=" is a SAFECHAR and should be allowed in unquoted values, so this is indeed a bug.
> Only comment I have is that passing in maxsplit=0 doesn't return the expected result (no splits). It's a sufficiently silly case that you can choose not to support it.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/collective/icalendar/pull/207#issuecomment-273746030

untitaker commented:
Also Jenkins is utterly broken and emails me because of that.

On Thu, Jan 19, 2017 at 03:03:28AM -0800, Lennart Regebro wrote:
> "=" is a SAFECHAR and should be allowed in unquoted values, so this is indeed a bug.
> Only comment I have is that passing in maxsplit=0 doesn't return the expected result (no splits). It's a sufficiently silly case that you can choose not to support it.
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/collective/icalendar/pull/207#issuecomment-273746030

regebro commented:
Well, maxsplit=0 should reasonably not split at all. maxsplit=0 being the same as maxsplit=1 is even more confusing than it being the same as maxsplit=-1. :-)

untitaker commented:
`maxsplit=0` returns a list with len <= 1, `maxsplit=1` a list with len <= 2, just like `str.split` does.

untitaker commented:
Ah, I see it fails in cases where there's nothing between the separator.

untitaker commented:
@regebro so that was a bug, fixed!

geier commented on src/icalendar/tests/test_icalendar.py:
```diff
@@ -247,7 +275,46 @@ def test_q_split(self):
+    def test_q_split_bin(self):
+        from ..parser import q_split
+        for s in ('X-SOMETHING=ABCDE==', ',,,'):
+            for maxsplit in range(3):
```
I'd suggest `range(-1, 3)` here.

untitaker replied:
done

stlaz commented:
I hope I am not missing anything but is the `q_split()` function really needed? Why not either use the Python's native `str.split()` or if we really really don't like the "." notation, map it in a function like this:
```python
def q_split(st, sep=",", max_split=0):
    st.split(sep, max_split)
```
The `q_split()` function looks like a hack from Python-prehistoric times to me.

geier commented:
Because it doesn't split if we are in between "s.

Quoting Stanislav Láznička (2017-01-19 18:49:27)
> I hope I am not missing anything but is the `q_split()` function really needed? Why not either use the Python's native `str.split()` or if we really really don't like the "." notation, map it in a function like this:
> ```python
> def q_split(st, sep=",", max_split=0):
>     st.split(sep, max_split)
> ```
> The `q_split()` function looks like a hack from Python-prehistoric times to me.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/collective/icalendar/pull/207#issuecomment-273847258

stlaz commented:
@geier Thank you, I found that out right after I wrote that comment so I removed it :) Been in work for too long today.

regebro commented:
Ah, pft, I apparently didn't see the initial "if maxsplit==0:" check. All good.

untitaker commented on src/icalendar/tests/test_icalendar.py:
```diff
@@ -247,7 +275,46 @@ def test_q_split(self):
+        Test if we support base64 encoded binary data in parameter values.
+        """
+        directory = os.path.dirname(__file__)
+        data = open(os.path.join(directory, 'x_location.ics'), 'rb').read()
```
Please use a context manager here to fix the resource warning.

geier replied:
That's the way it's done throughout the test suite. I believe it's better to be consistent than to be right (in this case).

untitaker replied:
It's no longer done this way, see #213

geier replied:
oh, it isn't any more... alright, will fix

untitaker commented:
This conflicts with master, otherwise LGTM

geier commented:
done

geier commented:
I can force push into the branch, and your approval doesn't get auto-revoked!?

geier commented:
anybody got any idea why python 2.6 now fails on travis with

    import unittest2 as unittest
    E   ImportError: No module named unittest2
?

geier commented:
As this group maintained, I'm not going to merge this myself, but I'd be grateful if somebody else would do it or voice their concerns.

I believe the current failure with python2.6 can safely be ignored, as it fails on master, too :grimacing:

hobarrera commented:
Any update on this? Looks ready, right?

untitaker commented:
@regebro could you take a quick look at this again?

geier commented:
Thanks everyone who contributed!
